features/netsocket/SocketAddress.cpp

```diff
@@ -215,17 +215,17 @@ void SocketAddress::set_port(uint16_t port)
 
 const char *SocketAddress::get_ip_address() const
 {
-    char *ip_address = (char *)_ip_address;
-
-    if (!ip_address[0]) {
+    if (!_ip_address[0]) {
         if (_addr.version == NSAPI_IPv4) {
-            ipv4_to_address(ip_address, _addr.bytes);
+            ipv4_to_address(_ip_address, _addr.bytes);
         } else if (_addr.version == NSAPI_IPv6) {
-            ipv6_to_address(ip_address, _addr.bytes);
+            ipv6_to_address(_ip_address, _addr.bytes);
+        } else {
```

Comment:
Not sure that compatibility with current code or being parsable as such is actually desirable. If you pass "unspec" to someone, do you really want them to silently treat it as if it was IPv4?

It kind of extends the somewhat undesirable behaviour of treating IPv4 0.0.0.0 as NULL and vice versa. (Nanostack had to have a special check to permit IPv4 iff it was 0.0.0.0...)  I'd prefer it if binary->text->binary round-tripping preserved value+type. (The stack doesn't do that as much as it used to, but it still can happen).

An unspec SocketAddress is only appropriate to a few quite limited cases - error returns, and bind are all I can currently think of.